Update placeholder for course outline block to include blank or AI option

donnapep opened the pull request:
Resolves #6956.

## Proposed Changes
Introduces a `window.sensei.aiCourseOutline` feature flag. If enabled, it shows the new UI:

![Screenshot 2023-06-14 at 12 03 56 PM](https://github.com/Automattic/sensei/assets/1190420/bc12e197-11b6-41a5-86bc-2bd22352c122)

Is disabled, it shows the old UI:

![Screenshot 2023-06-14 at 12 04 38 PM](https://github.com/Automattic/sensei/assets/1190420/c972ea6a-6041-49cd-9ced-2c924eb6b9b1)

## Testing Instructions

1. Run `npm run start` or `npm run build` for Sensei LMS.
2. Create a new course.
3. Step through the wizard and select the default layout.
4. Remove all lessons from the course outline in order to see the placeholder.
5. Check that the UI matches the [design](https://www.figma.com/file/NyhJy9d7SwvDoIAXecN7eH/Course-Creation?type=design&node-id=4352-33920&t=EGtFWFV4yaT8Efud-0).
6. Test it at different resolutions. At resolutions < 600px, the options should stack.
7. Set [this line](https://github.com/Automattic/sensei/compare/update/course-outline-placeholder-options?expand=1#diff-a37580803b282584a79424a39affab28b2d702ece20c9d4382d61cbb8c598b18R129) to `false` to disable the feature flag.
8. Refresh the browser and check that the old UI is visible.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [x] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [x] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [x] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [x] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [x] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

donnapep committed:
Add image assets

donnapep committed:
Add component for Sensei Pro badge

donnapep committed:
Add AI icon as component

donnapep committed:
Add feature flag

donnapep committed:
Update placeholder UI if feature flag is set

donnapep committed:
Add changelog entry

donnapep committed:
Merge branch 'trunk' into update/course-outline-placeholder-options

donnapep committed:
Add tests for <OutlinePlaceholder />

donnapep committed:
FIx placeholder item margins on WP 6.0

gabrielcaires commented on assets/blocks/course-outline/outline-block/outline-placeholder.js:
```diff
@@ -8,38 +8,118 @@ import { __ } from '@wordpress/i18n';
+				'sensei-lms'
+		  );
+
+	const content = window.sensei.aiCourseOutline ? (
```
Can we have it in a separate component? It is a bit to read inline components when they are so big.

donnapep replied:
I thought about that, but given that this code isn't going to be long-lived in its current state, I opted not to do that. If you feel strongly about it, I can do that though.

gabrielcaires commented on includes/blocks/class-sensei-course-blocks.php:
```diff
@@ -123,6 +123,13 @@ public function enqueue_block_editor_assets() {
+		wp_add_inline_script(
+			'sensei-single-course-blocks',
+			'window.sensei = window.sensei || {}; ' .
+			'window.sensei.aiCourseOutline = true;',
```
We need to have it as false by default in order to merge it on the trunk

donnapep replied:
😳 Fixed in https://github.com/Automattic/sensei/pull/6974/commits/08a9a0a7fae28488244b2ef33c2bfc2d7ccfec08.

gabrielcaires reviewed (changes requested):
I found an block and an opportunity to improve;.

Blocker: We need to keep the feature flag state as off by default. 
Bug: When we add the course outline block to a course with lessons, it first render the new screen and them render the existent lessons.  We probably need wait until load the existent lessons and only when they are empty we should show the new screen. 





If I add the course outline block to a course with lessons it shows the new screen and them show back the

donnapep committed:
Feature flag should be disabled by default!

gabrielcaires reviewed (dismissed):
I also found small layout incosistencies.

1) This distance should be 16px
<img width="677" alt="Screenshot 2023-06-15 at 15 04 45" src="https://github.com/Automattic/sensei/assets/38718/0edb48ab-8832-4bf4-a302-6ac849c86332">

2) We are missing the ✅  instead of the points on the bullets.
<img width="304" alt="Screenshot 2023-06-15 at 15 05 45" src="https://github.com/Automattic/sensei/assets/38718/8258dec0-cba3-498a-b62f-d8e9a03175ad">

3) The lessons should be text instead of images because they need to be translatable and because the text seems very blurred due the image compression. 
 

<img width="279" alt="Screenshot 2023-06-15 at 15 10 44" src="https://github.com/Automattic/sensei/assets/38718/e0c913c5-6d2f-47b5-9c5c-f299fcc0b762">

gabrielcaires commented on assets/blocks/course-outline/outline-block/outline-placeholder.test.js:
```diff
@@ -0,0 +1,48 @@
+			)
+		).toBeTruthy();
+		expect( getByText( 'Create a module' ) ).toBeTruthy();
+		expect( getByText( 'Create a lesson' ) ).toBeTruthy();
```
Is it duplicated?

donnapep replied:
Nope. These are separate buttons.

gabrielcaires replied:
Oh, I need my eyeglasses back 👓

donnapep committed:
Merge branch 'trunk' into update/course-outline-placeholder-options

donnapep commented:
> This distance should be 16px

I don't think we need to be pixel perfect here, do we? As with the LM project, it doesn't seem like a good use of time.

> The lessons should be text instead of images because they need to be translatable and because the text seems very blurred due the image compression.

Hmm, yes, I think you're right. Let me switch this one from an image as it will be strange otherwise for non-English sites.

donnapep commented:
> Bug: When we add the course outline block to a course with lessons, it first render the new screen and them render the existent lessons. We probably need wait until load the existent lessons and only when they are empty we should show the new screen.

I'm struggling with this one a bit. When I add the block to a course with lessons, the call to `getBlocks` returns an empty array until, eventually, it returns the blocks. For a course without any lessons, `getBlocks` will always return an empty array. 
So great, I need to detect when the call has actually resolved so that I can be confident that `getBlocks` really is an empty array.

However, adding a check like `isResolving: select( blockEditorStore ).isResolving( 'getBlocks', [ clientId ] ),` below [here](https://github.com/Automattic/sensei/blob/trunk/assets/blocks/course-outline/outline-block/outline-edit.js#L57) to see when the call has resolved always returns `false`. I also tried `hasFinishedResolution` with similar results (always the same value).

Can you see if I'm missing an obvious solution here?

donnapep commented:
> We are missing the ✅ instead of the points on the bullets.

It's a bit weird that only the AI option has checks. I wonder if either both (or none) should have them. Perhaps we should check with Andrei to see if that was intentional?

donnapep committed:
Fix styling issues on Astra

Move _placeholder-item inside placeholder-items selector and add editor-styles-wrapper class

gabrielcaires commented:
> > This distance should be 16px
> 
> I don't think we need to be pixel-perfect here, do we? As with the LM project, it doesn't seem like a good use of time.

Well, when the cost to do it is small (small === no variation by theme or no fluid) so we can be. Do we have something that makes it aligned hard to do? 


> It's a bit weird that only the AI option has been checked. I wonder if either both (or none) should have them. Let's check with Andrei to see if that was intentional.

 Checking here p1686849460494139-slack-C02NWDZBL0H

donnapep committed:
Don't use image for blank option so that text can be translated

donnapep committed:
Use checkmarks for AI list

donnapep committed:
CSS tweak for Divi

donnapep commented:
> Well, when the cost to do it is small (small === no variation by theme or no fluid) so we can be. Do we have something that makes it aligned hard to do?

The time cost is never small when you need to worry about multiple themes. 😢  I fixed this particular one [here](https://github.com/Automattic/sensei/pull/6974/commits/4a9e4802092b6d735fd696100ddecdb52634029a) though. In general, I'm trying to make the Course theme look _great_ and Astra and Divi to look _reasonable_.

donnapep commented:
@gabrielcaires [This](https://github.com/Automattic/sensei/pull/6974#issuecomment-1593385998) is the only outstanding item. I believe I've addressed the rest of the feedback. I can take another look at that last issue tomorrow, but I think it's also a bit of an edge case. I was only able to reproduce it if I removed the block, saved, then added it back. But not if I removed the block, saved, refreshed the page, and then added the block again.

m1r0 commented:
Here are some things that I've noticed related to the html-vs-design:

* The block icon is different. Not sure if this is out of scope.

Current:
<img width="206" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/49e5c50b-49a3-42d6-9394-4c633297e140">

Design:
<img width="194" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/b8a46346-4321-4d10-bfc3-046867b84a1b">

* The placeholder item has a 2px border-radius in the design.

<img width="50" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/a33a198c-d45a-4d5f-b5fc-2f897b846941">

* The hat icon is a bit cut off.

Current:
<img width="30" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/a9f0fd17-e3cd-453f-8fb3-325abf358b04">

Design:
<img width="43" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/008354ec-e5be-4774-8b57-ee4ced1263ee">

* The placeholder item image is not that sharp on my screen (pixelated). Maybe use a 2x image?

<img width="322" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/f6ef0b2b-99bb-4afc-96bf-a7135d4bf8a9">

* At some places in the design the "No starting point help" text is "Add Lessons and Modules". Same for "General course outline" -> "Reorder and edit anytime". Not sure which is the correct one.

donnapep committed:
Update AI icon and add new AI loader icon

donnapep committed:
Replace PNG with SVG

donnapep committed:
Update text for blank option

donnapep commented:
> The block icon is different. Not sure if this is out of scope.

I think this one should be a new card since this is more about updating the block icon, which will update this UI automatically. @gabrielcaires 

> The placeholder item has a 2px border-radius in the design.

Fixed as part of https://github.com/Automattic/sensei/pull/6974/commits/7122317e54d493b9e5402a610fea69852b9e8093.

> The hat icon is a bit cut off.

This icon has been replaced with a new icon in https://github.com/Automattic/sensei/pull/6974/commits/7122317e54d493b9e5402a610fea69852b9e8093.

> The placeholder item image is not that sharp on my screen (pixelated). Maybe use a 2x image?

Switched to an SVG in https://github.com/Automattic/sensei/pull/6974/commits/fec8a1023f420b0688f3ec6d532394a06100e776.

> At some places in the design the "No starting point help" text is "Add Lessons and Modules". Same for "General course outline" -> "Reorder and edit anytime". Not sure which is the correct one.

I think the design has been updated since I created the UI, so I've updated it to the new text in https://github.com/Automattic/sensei/pull/6974/commits/9cb90aab5fadccf7f863770cdddb56bc166aad64.

m1r0 commented:
Codewise, it looks solid, but I've noticed a few more design misalignments:

* The `item-details` text between the left and right items looks off. On the design, they look on the same row. This also offsets the image below.

Current:
<img width="946" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/cc68bde5-a4d0-4e18-abc2-f79b58b470f1">

Design:
<img width="722" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/bbbab6fe-21a1-45ad-a062-ab361165e6f3">

* On the smaller mobile resolutions, the items go outside of the container. You can ignore this if we don't support them.

<img width="466" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/0370cdca-a185-4e4f-b861-7c350ac5c06a">


* On ipad/some mobile landspace, when the sidebar is extended, the placeholder items have different widths. Also, the right item goes outside of the container.


<img width="844" alt="image" src="https://github.com/Automattic/sensei/assets/1612178/b1922e0a-4f23-4505-90f3-828e3083041f">

gabrielcaires committed:
Merge branch 'trunk' into update/course-outline-placeholder-options

gabrielcaires commented:
@m1r0 @donnapep  Are you ok with move the last layout adjustments to another PR or card? I need this code on my card

m1r0 commented:
@gabrielcaires Yep, sounds good to me. 👍
